sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -195,7 +228,7 @@ trait SpillableHandle extends StoreHandle {
 trait DeviceSpillableHandle[T <: AutoCloseable] extends SpillableHandle {
   private[spill] var dev: Option[T]
 
-  private[spill] override def spillable: Boolean = synchronized {
+  private[spill] override def spillable: Boolean = {
     super.spillable && dev.isDefined
```

Comment:
`super.spillable` contains a lock now, meaning if it is true then nobody else is currently spilling, which means `dev` could be defined or not but it will be in a consistent state. I could be missing something though.

Reply:
well, the call to `super.spillable` here returns and then we check `dev` outside of the lock. You could have another thread that started spilling after `super.spillable` so the state of `dev` is not really consistent here, but I don't think it matters. The spiller thread looks at this result and either includes or excludes the handle in its plan. As long as a thread has included the handle in their plan that's good enough, someone is going to spill it. Yes, lets update the doc at the top of the file with this behavior.

Reply:
actually it does matter I think. I don't think setting an `Option[X]` or calling methods on one, are thread safe. We may need to mark it `volatile` or synchronize around the check of the variable, since it is getting updated from different threads.

Reply:
Ok thinking more about it, maybe this is sufficient:

```
private[spill] override def spillable: Boolean = {
    if (super.spillable) {
      synchronized {
        dev.getOrElse {
          throw new IllegalStateException(
            s"$this is spillable but it doesn't have a dev buffer!")
        }.getRefCount == 1
      }
    } else {
      false
    }
  }
```
Then, even if another thread starts spilling, we will get a consistent view of `dev` - but we still have the ability to bail early in most cases if another thread is already spilling. In the race condition described above, we could still end up getting blocked on IO which this PR aims to removed, but it would be very rare and safe at least.

Reply:
I also considered introducing another dedicated lock for `dev` but I think it still reduces to the same problem and further complicates things